PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/common/AzureArtifactComboBox.java

```diff
@@ -0,0 +1,147 @@
+        this.setLoading(true);
+        subscription = this.loadItemsAsync()
+                           .subscribe(items -> DefaultLoader.getIdeHelper().invokeLater(() -> {
+                               this.removeAllItems();
+                               this.setItems(items);
+                               this.setLoading(false);
+                               this.resetDefaultValue(defaultArtifactType, defaultPath, artifactIdentifier);
+                           }), (e) -> {
+                                   this.handleLoadingError(e);
+                               });
+    }
+
+    @NotNull
+    @Override
+    protected List<? extends AzureArtifact> loadItems() throws Exception {
+        return AzureArtifactManager.getInstance(project).getAllSupportedAzureArtifacts();
+    }
+
+    @Nullable
+    @Override
+    protected ExtendableTextComponent.Extension getExtension() {
+        return ExtendableTextComponent.Extension.create(
+                AllIcons.General.OpenDisk, "Open file", this::onSelectFile);
+    }
+
+    protected String getItemText(Object item) {
+        if (item instanceof AzureArtifact) {
+            return String.format("%s : %s", ((AzureArtifact) item).getType(), ((AzureArtifact) item).getName());
+        } else {
+            return StringUtils.EMPTY;
+        }
+    }
+
+    @Nullable
+    protected Icon getItemIcon(Object item) {
+        return item instanceof AzureArtifact ? ((AzureArtifact) item).getIcon() : null;
+    }
+
+    private void onSelectFile() {
+        final FileChooserDescriptor fileDescriptor = FileChooserDescriptorFactory.createSingleFileDescriptor();
+        if (filter != null) {
+            fileDescriptor.withFileFilter(filter);
+        }
+        fileDescriptor.withTitle("Select artifact to deploy");
+        final VirtualFile file = FileChooser.chooseFile(fileDescriptor, null, null);
+        if (file != null && file.exists()) {
+            addOrSelectExistingVirtualFile(file);
+        }
+    }
+
+    private void addOrSelectExistingVirtualFile(VirtualFile virtualFile) {
+        final AzureArtifact selectArtifact = AzureArtifact.createFromFile(virtualFile);
+        final List<AzureArtifact> artifacts = UIUtils.listComboBoxItems(AzureArtifactComboBox.this);
+        final AzureArtifact existingArtifact =
+                artifacts.stream().filter(artifact -> artifact.getType() == AzureArtifactType.Artifact
+                        && StringUtils.equalsAnyIgnoreCase(artifact.getTargetPath(), selectArtifact.getTargetPath()))
+                         .findFirst().orElse(null);
+        if (existingArtifact == null) {
+            this.addItem(selectArtifact);
+            this.setSelectedItem(selectArtifact);
+        } else {
+            this.setSelectedItem(existingArtifact);
+        }
+    }
+
+    private void resetDefaultValue(final AzureArtifactType defaultArtifactType, final String defaultPath, final String artifactIdentifier) {
```

Comment:
seems very similar to super's `setValue` method.

Reply:
Will refactor in later PR